### PR TITLE
 Purge queues during reindexing (#1185) 

### PR DESCRIPTION
**Pull request by jessebrennan**
An outstanding question: should `--purge` default to True?

### CODE REVIEW

**Comment by achave11** [comment]
Regarding your question. I think it should be default. Purged queues allow a fresh reindex to have a clean slate. In regard to the management of those message I am not sure.

**Comment by jessebrennan** [comment]
@achave11 I don't understand the last part of your comment. To stick with the conventions of the other options in `reindex.py`, I'm going to let `--purge` default to `False`, but I'm curious what @hannes-ucsc  thinks.

File: scripts/reindex.py [hunk 1]
```diff
@@ -80,6 +81,11 @@ def __init__(self, prog) -> None:
                     default=False,
                     action='store_true',
                     help='Index all matching bundles in the configured DSS instance.')
+parser.add_argument('--purge',
+                    default=False,
+                    action='store_true',
+                    help='Freeze the lambdas and purge the queues before reindexing. On success, lambdas are'
```
Comment (hannes-ucsc) [inline scripts/reindex.py]:
```suggestion
                    help='Purge the queues before taking any actions on the index.')
```

The lambda enabling/disabling is an implementation detail of the purge action.

File: scripts/reindex.py [hunk 2]
```diff
@@ -80,6 +81,11 @@ def __init__(self, prog) -> None:
...
+                         'reactivated.')
```
Comment (hannes-ucsc) [inline scripts/reindex.py]:
```suggestion
```

File: scripts/reindex.py [hunk 3]
```diff
@@ -104,9 +110,22 @@ def main(argv: List[str]):
...
+        queue_manager.purge_queues(queues)
     if args.delete:
+        logger.info('Deleting indices...')
         azul_client.delete_all_indices()
+    if args.purge:
+        logger.info('Reenabling lambdas...')
+        # noinspection PyUnboundLocalVariable
+        queue_manager.manage_lambdas(queues, enable=True)
     if args.index:
+        logger.info('Queuing reindex notifications...')
```
Comment (hannes-ucsc) [inline scripts/reindex.py]:
```suggestion
        logger.info('Queuing notifications for reindexing...')
```

File: src/azul/queues.py [hunk 4]
```diff
@@ -189,17 +189,20 @@ def _cleanup():
...
-        self._purge_queues({queue_name: queue})
+        self.purge_queues_while_disabled({queue_name: queue})
 
     def purge_all(self):
-        self._purge_queues(dict(self._azul_queues()))
+        self.purge_queues_while_disabled(dict(self.azul_queues()))
 
-    def _purge_queues(self, queues: MutableMapping[str, Queue]):
-        self._manage_lambdas(queues, enable=False)
+    def purge_queues_while_disabled(self, queues: MutableMapping[str, Queue]):
```
Comment (hannes-ucsc) [inline src/azul/queues.py]:
```suggestion
    def purge_queues_safely(self, queues: MutableMapping[str, Queue]):
        """
        Disable lambdas, purge the given queues and re-enable the lambdas.
```

File: src/azul/queues.py [hunk 5]
```diff
@@ -189,17 +189,20 @@ def _cleanup():
...
+        self.manage_lambdas(queues, enable=False)
+        self.purge_queues(queues)
+        self.manage_lambdas(queues, enable=True)
+
+    def purge_queues(self, queues: MutableMapping[str, Queue]):
```
Comment (hannes-ucsc) [inline src/azul/queues.py]:
```suggestion
    def purge_queues_unsafely(self, queues: MutableMapping[str, Queue]):
        """
        Purge the given queues without disabling the lambdas.
```

**Comment by hannes-ucsc** [comment]
> An outstanding question: should `--purge` default to True?

Leaning towards no.

**Pull request merged**